### PR TITLE
RL Chain with VowpalWabbit

### PR DESCRIPTION
**Pull request by olgavrou**
- Description: This PR adds a new chain `rl_chain.PickBest` for learned prompt variable injection, detailed description and usage can be found in the example notebook added. It essentially adds a [VowpalWabbit](https://github.com/VowpalWabbit/vowpal_wabbit) layer before the llm call in order to learn or personalize prompt variable selections.

Most of the code is to make the API simple and provide lots of defaults and data wrangling that is needed to use Vowpal Wabbit, so that the user of the chain doesn't have to worry about it.

  - Dependencies: [vowpal-wabbit-next](https://pypi.org/project/vowpal-wabbit-next/),
     - sentence-transformers (already a dep)
     - numpy (already a dep)
  - tagging @ataymano who contributed to this chain
  - Tag maintainer: @baskaryan
  - Twitter handle: @olgavrou

Added example notebook and unit tests

### CODE REVIEW

**Comment by baskaryan** [comment]
this is super cool @olgavrou! i think it probably makes more sense for this to live in langchain_experimental (`libs/experimental/langchain_experimental`) package than langchain package. we're trying to put all new more end-to-end chains in langchain_experimental and keeping only the basic, building-block chains in langchain. the difference in user experience should be small. if that sounds good to you im happy to make the move!

**Comment by olgavrou** [comment]
Thanks @baskaryan that sounds good 

I can also see that the 3.11 extended test fails due to a torchvision dependency installation. I’m unsure why this is happening, do you have any ideas maybe?

**Comment by baskaryan** [comment]
> Thanks @baskaryan that sounds good
> 
> I can also see that the 3.11 extended test fails due to a torchvision dependency installation. I’m unsure why this is happening, do you have any ideas maybe?

I believe there are some known issues with poetry and torchvision https://github.com/python-poetry/poetry/issues/6409. if we don't want to let this block things we can move the unit tests to integration tests for now?

**Comment by olgavrou** [comment]
> > Thanks @baskaryan that sounds good
> > I can also see that the 3.11 extended test fails due to a torchvision dependency installation. I’m unsure why this is happening, do you have any ideas maybe?
> 
> I believe there are some known issues with poetry and torchvision [python-poetry/poetry#6409](https://github.com/python-poetry/poetry/issues/6409). if we don't want to let this block things we can move the unit tests to integration tests for now?

Sure, I can make both changes: moving into experimental and moving unit to integration tests

**Comment by olgavrou** [comment]
> Sure, I can make both changes: moving into experimental and moving unit to integration tests

ok @baskaryan moved things into langchain_experimental, leaving the tests as `unit_tests` seems to work fine now.

Two things:
 1. Unrelated, there were some unit tests failing in experimental, looked unrelated to my changes, added a workaround, let me know if you think it is ok
2. I also left the notebook in the original location, let me know if there is a different location for it that would make more sense for the docs, like under `use cases` maybe

**Comment by olgavrou** [comment]
@baskaryan :)

File: libs/experimental/pyproject.toml [hunk 1]
```diff
@@ -14,6 +14,9 @@ langchain = ">=0.0.239"
 presidio-anonymizer = {version = "^2.2.33", optional = true}
 presidio-analyzer = {version = "^2.2.33", optional = true}
 faker = {version = "^19.3.1", optional = true}
+vowpal-wabbit-next = {version = "0.6.0", optional = true}
+sentence-transformers = {version = "^2", optional = true}
+pandas = {version = "^2.0.1", optional = true}
```
Comment (baskaryan) [inline libs/experimental/pyproject.toml:19]: is pandas needed for unit tests? if not could we drop it here. think any dep that's not in an extras group will be default installed, even if it's optional.

Comment (olgavrou) [reply libs/experimental/pyproject.toml:19]: @baskaryan it isn't needed for the unit tests but it is needed for the chain to work. If I remove the pandas dep from here and re-build/install langchain_experimental, then the notebook I have breaks, so not sure I understand where pandas will become installed from

Comment (olgavrou) [reply libs/experimental/pyproject.toml:19]: @baskaryan tagging again :)

Comment (olgavrou) [reply libs/experimental/pyproject.toml:19]: @baskaryan I guess I'll tag once again :)

Comment (baskaryan) [reply libs/experimental/pyproject.toml:19]: user would be expected to manually install themselves, same as with all optional dependencies

**Comment by baskaryan** [comment]
apologies for the delay, merged into master in #11462! thank you @olgavrou and @ataymano!!

**Pull request merged**